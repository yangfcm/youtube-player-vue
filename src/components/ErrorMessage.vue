<script setup lang="ts">
defineProps<{
  message?: string
}>()
</script>

<template>
  <div>
    <v-alert icon="mdi-alert" type="error">{{ message || 'Oops, something went wrong.' }}</v-alert>
  </div>
</template>
