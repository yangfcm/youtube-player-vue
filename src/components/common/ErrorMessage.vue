<template>
  <div>
    <app-blank></app-blank>
    <h3 class="ui red header center aligned">
      <i class="exclamation triangle icon"> </i><slot></slot>
    </h3>
    <app-blank></app-blank>
  </div>
</template>

<script>
import Blank from "./Blank";
export default {
  components: {
    appBlank: Blank,
  },
  props: ["errorMessage"],
};
</script>
<style scoped></style>
