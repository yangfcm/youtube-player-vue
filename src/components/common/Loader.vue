<template>
  <div>
    <app-blank></app-blank>
    <div class="ui active centered inline loader"></div>
    <app-blank></app-blank>
  </div>
</template>

<script>
import Blank from "./Blank";
export default {
  components: {
    appBlank: Blank,
  },
};
</script>

<style scoped></style>
