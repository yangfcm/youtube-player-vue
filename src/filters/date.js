/**
 * Convert a date object to human-readable string
 */
export default (date) => {
  const dateObj = new Date(date);
  const day = dateObj.getDate();
  const month = new Intl.DateTimeFormat("en-US", {
    month: "short",
  }).format(dateObj);
  const year = dateObj.getFullYear();
  return `${day} ${month}, ${year}`;
};
