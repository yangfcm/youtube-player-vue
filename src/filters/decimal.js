/**
 * Separate an integer number by commas
 * e.g. 12345678 => 12,345,678
 */
export default (num) => {
  if (isNaN(Number(num))) return "";
  let strNum = String(num);
  let arrNum = strNum.split("");
  let arr = [];
  let part = "";
  for (let i = 0; i < arrNum.length; i++) {
    let num = arrNum[arrNum.length - i - 1];
    part = num + part;
    if ((i + 1) % 3 === 0) {
      arr.unshift(part);
      part = "";
    }
  }
  if (part.length > 0) {
    arr.unshift(part);
  }
  return arr.join(",");
};
