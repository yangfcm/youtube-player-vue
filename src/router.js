import Vue from "vue";
import VueRouter from "vue-router";
import Home from "./components/pages/Home";

const MyChannel = () => import("./components/pages/MyChannel.vue");
const MyPlaylist = () => import("./components/pages/MyPlaylist.vue");
const Video = () => import("./components/pages/Video.vue");
const PlaylistDetail = () => import("./components/pages/PlaylistDetail.vue");
const Channel = () => import("./components/pages/Channel.vue");
const SearchResult = () => import("./components/pages/SearchResult.vue");
const ChannelVideos = () => import("./components/pages/ChannelVideos.vue");
const ChannelPlaylist = () => import("./components/pages/ChannelPlaylist.vue");
const ChannelIntro = () => import("./components/pages/ChannelIntro.vue");
const NotFound = () => import("./components/pages/NotFound.vue");

const routes = [
  { path: "/", component: Home },
  { path: "/subscription", component: MyChannel },
  { path: "/playlist", component: MyPlaylist },
  { path: "/video/:id", component: Video },
  { path: "/playlist/:id", component: PlaylistDetail },
  {
    path: "/channel/:id",
    component: Channel,
    children: [
      { path: "", redirect: "/channel/:id/videos" },
      { path: "videos", component: ChannelVideos },
      {
        path: "playlist",
        component: ChannelPlaylist,
      },
      {
        path: "intro",
        component: ChannelIntro,
      },
    ],
  },
  { path: "/search/:keyword", component: SearchResult },
  { path: "*", component: NotFound },
];

Vue.use(VueRouter);

export default new VueRouter({
  mode: "history",
  routes,
});
