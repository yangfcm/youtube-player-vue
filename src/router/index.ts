import { createRouter, createWebHistory } from 'vue-router'
import HomePage from '@/pages/HomePage.vue'
import ChannelVideos from '@/pages/ChannelVideos.vue'
import ChannelPlayLists from '@/pages/ChannelPlayLists.vue'
import NotFound from '@/pages/NotFound.vue'

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    {
      path: '/',
      name: 'home',
      component: HomePage,
    },
    {
      path: '/subscription',
      name: 'subscription',
      component: () => import('../pages/SubscriptionPage.vue'),
    },
    {
      path: '/playlist',
      name: 'playlist',
      component: () => import('../pages/PlayListPage.vue'),
    },
    {
      path: '/video/:id',
      name: 'video',
      component: () => import('../pages/VideoPage.vue'),
    },
    {
      path: '/playlist/:id',
      name: 'playlist-videos',
      component: () => import('../pages/PlayListVideos.vue'),
    },
    {
      path: '/search/:keyword',
      name: 'search-result',
      component: () => import('../pages/SearchResultsPage.vue'),
    },
    {
      path: '/channel/:id',
      name: 'channel',
      component: () => import('../pages/ChannelPage.vue'),
      children: [
        {
          path: '',
          name: 'channel-index',
          redirect: (to) => ({
            path: `/channel/${to.params.id}/videos`,
          }),
        },
        {
          path: 'videos',
          name: 'channel-videos',
          component: ChannelVideos,
        },
        {
          path: 'playlists',
          name: 'channel-playlists',
          component: ChannelPlayLists,
        },
      ],
    },
    {
      path: '/:pathMatch(.*)*',
      component: NotFound,
    },
  ],
})

export default router
