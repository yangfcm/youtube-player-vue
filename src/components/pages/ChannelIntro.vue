<template>
  <div>
    <app-loader v-if="!channelIntro && !error"></app-loader>
    <app-error-message v-if="error">{{ error }}</app-error-message>
    <div v-if="channelIntro && !error" class="ui segments" id="app-channel-intro">
      <div class="ui segment">
        <p>
          <i class="calendar plus icon"></i>
          Published on {{ channelIntro.snippet.publishedAt | date }}
        </p>
      </div>
      <div class="ui segment">
        <p>
          <i class="file alternate icon"></i>
          Description -
          {{ channelIntro.snippet.description }}
        </p>
      </div>
      <div class="ui segment">
        <p>
          <i class="users icon"></i>
          {{ channelIntro.statistics.subscriberCount | decimal }}
          subscribers
        </p>
      </div>
      <div class="ui segment">
        <p>
          <i class="eye icon"></i>
          {{ channelIntro.statistics.viewCount | decimal }} views
        </p>
      </div>
    </div>
  </div>
</template>

<script>
import { mapState, mapGetters, mapActions } from "vuex";
import ErrorMessage from "../common/ErrorMessage";
import Loader from "../common/Loader";

export default {
  components: {
    appLoader: Loader,
    appErrorMessage: ErrorMessage,
  },
  data: function () {
    return {
      channelIntro: null,
      error: "",
    };
  },
  computed: {
    ...mapState(["channel"]),
    ...mapGetters(["channelErrorMessage"]),
  },
  methods: {
    ...mapActions(["fetchChannelIntro"]),
  },
  async created() {
    const channelId = this.$route.params.id;
    await this.fetchChannelIntro(channelId);
    if (this.channel.channelError) {
      this.error = this.channelErrorMessage;
    } else if (this.channel.channelIntro) {
      this.channelIntro = this.channel.channelIntro.items[0];
      this.error = "";
      console.log(this.channelIntro);
      if (!this.channelIntro) {
        console.log("not found");
        this.$router.push("/not-found").catch(() => {
          console.log("error");
        });
      }
    }
  },
};
</script>

<style scoped></style>
