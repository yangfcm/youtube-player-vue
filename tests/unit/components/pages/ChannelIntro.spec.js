import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import VueRouter from "vue-router";
import { routes } from "@/router";
import ChannelIntro from "@/components/pages/ChannelIntro";
import Loader from "@/components/common/Loader";
import ErrorMessage from "@/components/common/ErrorMessage";
import { channelIntroResponse, channelId } from "../../store/fixtures/channel";
import date from "@/filters/date";
import decimal from "@/filters/decimal";

describe("Test ChannelIntro page", () => {
  const localVue = createLocalVue();
  localVue.use(Vuex);
  localVue.use(VueRouter);
  localVue.filter("date", date);
  localVue.filter("decimal", decimal);

  const router = new VueRouter({
    mode: "history",
    routes,
  });
  const actions = {
    fetchChannelIntro: jest.fn(() => {}),
  };
  const store = new Vuex.Store({
    state: {
      channel: {
        channelIntro: null,
        channelError: null,
      },
    },
    getters: {
      channelErrorMessage: jest.fn(() => ""),
    },
    actions,
  });
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(ChannelIntro, {
      store,
      localVue,
      router,
    });
  });
  afterEach(() => {
    wrapper.destroy();
  });

  it("should render Loader when no error or channel is returned", () => {
    expect(wrapper.findComponent(Loader).exists()).toBe(true);
    expect(wrapper.findComponent(ErrorMessage).exists()).toBe(false);
    expect(wrapper.find("div#app-channel-intro").exists()).toBe(false);
  });

  it("should get correct channelId and trigger action when component is initially rendered", async () => {
    router.push({
      name: "channelIntro",
      params: {
        id: channelId,
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$route.params.id).toBe(channelId);
    expect(actions.fetchChannelIntro).toHaveBeenCalled();
  });

  it("should render ErrorMessage component if error occurs", async () => {
    const errorMessage = "mock error message";
    wrapper.setData({
      error: errorMessage,
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.findComponent(Loader).exists()).toBe(false);
    expect(wrapper.findComponent(ErrorMessage).exists()).toBe(true);
    expect(wrapper.find("div#app-channel-intro").exists()).toBe(false);
    expect(wrapper.text()).toContain(errorMessage);
  });

  describe("Test Channel intro content if data is succssfully fetched", () => {
    beforeEach(() => {
      wrapper.setData({
        channelIntro: channelIntroResponse.items[0],
        error: "",
      });
    });

    it("should render div#app-channel-intro", () => {
      expect(wrapper.findComponent(Loader).exists()).toBe(false);
      expect(wrapper.findComponent(ErrorMessage).exists()).toBe(false);
      expect(wrapper.find("div#app-channel-intro").exists()).toBe(true);
    });

    it("channel intro content should match snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
