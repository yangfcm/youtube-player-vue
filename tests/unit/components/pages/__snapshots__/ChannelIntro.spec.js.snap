// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ChannelIntro page Test Channel intro content if data is succssfully fetched channel intro content should match snapshot 1`] = `
<div>
  <!---->
  <!---->
  <div id="app-channel-intro" class="ui segments">
    <div class="ui segment">
      <p><i class="calendar plus icon"></i>
        Published on 18 Jun, 2008
      </p>
    </div>
    <div class="ui segment">
      <p><i class="file alternate icon"></i>
        Description -
        Learn how to speak English with the BBC. Every day we have a new video to help you learn the English language. We also produce regular 'extra' videos across the week so come back every day to see what's new.

        We like receiving and reading your comments - please use English when you comment.

        For more videos and content that will help you learn English, visit our website: http://www.bbclearningenglish.com
      </p>
    </div>
    <div class="ui segment">
      <p><i class="users icon"></i>
        2,850,000
        subscribers
      </p>
    </div>
    <div class="ui segment">
      <p><i class="eye icon"></i>
        138,098,330 views
      </p>
    </div>
  </div>
</div>
`;
